File: src/page/support/AnimTableTH.java
package page.support;

import javax.swing.*;
import java.awt.*;
import java.awt.datatransfer.Transferable;
import java.awt.dnd.DragSource;

import static page.support.AnimTransfer.DFS;

public class AnimTableTH<T> extends TransferHandler {

    private static final long serialVersionUID = 1L;

    private final int type;

    private final AnimTable<T> table;

    private int[] rows;

    public AnimTableTH(AnimTable<T> jtable, int t) {
        table = jtable;
        table.setDragEnabled(true);
        table.setDropMode(DropMode.INSERT_ROWS);
        type = t;
    }

    @Override
    public boolean canImport(TransferSupport info) {
        boolean b = info.getComponent() == table;
        b &= info.isDataFlavorSupported(DFS[type]);
        table.setCursor(b ? DragSource.DefaultMoveDrop : DragSource.DefaultMoveNoDrop);
        return b;
    }

    @Override
    public int getSourceActions(JComponent c) {
        return TransferHandler.COPY_OR_MOVE;
    }

    @Override
    public boolean importData(TransferSupport info) {
        if (!canImport(info))
            return false;
        try {
            if (info.isDrop()) {
                int row = ((JTable.DropLocation) info.getDropLocation()).getRow();
                return table.reorder(row, rows);
            } else {
                @SuppressWarnings("unchecked")
                T[] data = (T[]) info.getTransferable().getTransferData(DFS[type]);
                int row = table.getSelectedRow() + 1;
                return table.insert(row, data);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    @Override
    protected Transferable createTransferable(JComponent c) {
        assert (c == table);
        rows = table.getSelectedRows();
        return new AnimTransfer<>(type, table.getSelected());
    }

    @Override
    protected void exportDone(JComponent c, Transferable t, int act) {
        table.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));

    }

}

File: src/page/support/AnimTableTH.kt
package page.support

import page.support.AnimTransfer
import java.awt.Cursor
import java.awt.datatransfer.Transferable
import java.awt.dnd.DragSource
import javax.swing.DropMode
import javax.swing.JComponent
import javax.swing.JTable
import javax.swing.TransferHandler

class AnimTableTH<T>(jtable: AnimTable<T>, t: Int) : TransferHandler() {
    private val type: Int
    private val table: AnimTable<T>
    private var rows: IntArray
    override fun canImport(info: TransferSupport): Boolean {
        var b = info.getComponent() === table
        b = b and info.isDataFlavorSupported(AnimTransfer.Companion.DFS.get(type))
        table.setCursor(if (b) DragSource.DefaultMoveDrop else DragSource.DefaultMoveNoDrop)
        return b
    }

    override fun getSourceActions(c: JComponent): Int {
        return TransferHandler.COPY_OR_MOVE
    }

    override fun importData(info: TransferSupport): Boolean {
        if (!canImport(info)) return false
        try {
            return if (info.isDrop()) {
                val row: Int = (info.getDropLocation() as JTable.DropLocation).getRow()
                table.reorder(row, rows)
            } else {
                val data = info.getTransferable().getTransferData(AnimTransfer.Companion.DFS.get(type)) as Array<T>
                val row: Int = table.getSelectedRow() + 1
                table.insert(row, data)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    protected override fun createTransferable(c: JComponent): Transferable {
        assert(c === table)
        rows = table.getSelectedRows()
        return AnimTransfer<T>(type, table.getSelected())
    }

    protected override fun exportDone(c: JComponent, t: Transferable, act: Int) {
        table.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR))
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        table = jtable
        table.setDragEnabled(true)
        table.setDropMode(DropMode.INSERT_ROWS)
        type = t
    }
}
